### PR TITLE
Upgrade bundled hwloc to 2.9

### PR DESCRIPTION
**Pull request by jhh67**
We've been using a very old version of `hwloc` (1.11.13). Upgrading gives us access to new `hwloc` features including synthetic topologies, which will be useful for unit-testing the runtime.

Resolves https://github.com/Cray/chapel-private/issues/4590 and resolves https://github.com/Cray/chapel-private/issues/3585 and resolves https://github.com/Cray/chapel-private/issues/4503 and resolves https://github.com/Cray/chapel-private/issues/795.

### CODE REVIEW

**Comment by jhh67** [comment]
Elliot and I have decided to wait to merge this until after the 1.31 release.

File: runtime/src/topo/hwloc/topo-hwloc.c [hunk 1]
```diff
@@ -144,27 +144,16 @@ void chpl_topo_init(void) {
...
-  // Require at least hwloc version 1.11 (we need 1.11.5 to not crash
-  // in some NUMA configurations).
-  // Check both at build time and run time.
-#define REQUIRE_HWLOC_VERSION 0x00010b00
-
-#if HWLOC_API_VERSION < REQUIRE_HWLOC_VERSION
-#error hwloc version 1.11.5 or newer is required
-#endif
-
-  CHK_ERR(hwloc_get_api_version() >= REQUIRE_HWLOC_VERSION);
```
Comment (ronawho) [inline runtime/src/topo/hwloc/topo-hwloc.c:157]: Should we check if the version is >= 2 now?

File: third-party/hwloc/README [hunk 2]
```diff
@@ -18,16 +18,25 @@ The directory $CHPL_HOME/third-party/hwloc/hwloc-src/ contains the
...
 If the hwloc API changes, the runtime shim may need to be changed and
 incompatible versions need to be dropped from the whitelist.
+
+Chapel modifications to hwloc
+================================
+
+We made the following modifications to the official hwloc release:
+
+* Add HWLOC_DARWIN_LDFLAGS to Libs.private in hwloc.pc.in so that
+Chapel programs link correctly on a Mac.
```
Comment (ronawho) [inline third-party/hwloc/README:41]: Please open an hwloc issue for this (ideally a standalone reproducer that doesn't invovle chapel)

File: third-party/hwloc/README [hunk 3]
```diff
@@ -18,16 +18,25 @@ The directory $CHPL_HOME/third-party/hwloc/hwloc-src/ contains the
...
 5. update the version number mentioned above
 6. update the okversions whitelist in chpl_hwloc.py
 7. make sure these instructions are up to date :)
 8. test (std config, and that stream gets expected performance)
 9. commit, PR, merge, etc
 
-Note that these instructions are for simple 1.X API compatible updates.
+Note that these instructions are for simple 2.X API compatible updates.
 If the hwloc API changes, the runtime shim may need to be changed and
 incompatible versions need to be dropped from the whitelist.
```
Comment (ronawho) [inline third-party/hwloc/README:33]: This and the above numbered instructions need to be updated to remove the whitelist instructions

File: third-party/libfabric/Makefile [hunk 4]
```diff
@@ -17,8 +17,8 @@ endif
 CHPL_LIBFABRIC_CFG_OPTIONS += --enable-static \
                               --disable-shared \
                               --prefix=$(LIBFABRIC_INSTALL_DIR) \
-			      --disable-psm3 \
-			      --disable-opx
+                              --disable-psm3 \
+                              --disable-opx
```
Comment (ronawho) [inline third-party/libfabric/Makefile:21]: Minor, but let's not include unrelated changes in this already large PR

File: util/chplenv/third_party_utils.py [hunk 5]
```diff
@@ -178,6 +178,25 @@ def pkgconfig_get_system_link_args(pkg, static=pkgconfig_default_static()):
...
+    i = 0
+    while (i < len(args)):
+        if args[i] == '-framework':
+            result.append(args[i] + " " + args[i+1])
+            i += 2
+        else:
+            result.append(args[i])
+            i += 1
+    return result
+
```
Comment (ronawho) [inline util/chplenv/third_party_utils.py:199]: This feels fragile. I'd prefer that we instead update the `dedup` to not remove `-framework` or maybe to only dedup `-l` args.

**Review by ronawho — COMMENTED** [review]
Also resolves https://github.com/Cray/chapel-private/issues/795

Please check:
 - That stream performance looks good to verify pinning/affinity is correct
 - Check no-op startup time since hwloc is responsible for most of the startup time
 - Something with using multiple NICs to verify that's still working

**Comment by jhh67** [comment]
Stream performance results collected on an HPE Cray EX. `stream-main` was compiled using the Chapel `main` branch, `stream` was compiled using this PR.

```
> chpl --fast test/release/examples/benchmarks/hpcc/stream.chpl

>  ./stream-main -v -nl 16
...
Execution time:
  tot = 8.55375
  avg = 0.855375
  min = 0.84249
Performance (GB/s) = 2562.63

> ./stream -v -nl 16
...
Execution time:
  tot = 8.42603
  avg = 0.842603
  min = 0.827771
Performance (GB/s) = 2608.2

**Comment by ronawho** [comment]
EX uses a fixed heap and all pages are interleaved ,so even if no pinning was occurring I wouldn't expect perf to be much different. I'd recommend using comm=none

**Comment by jhh67** [comment]
Stream performance results collected on an HPE Cray EX with `CHPL_COMM=none`. `stream-main` was compiled using the Chapel `main` branch, `stream` was compiled using this PR.
```
> chpl --fast test/release/examples/benchmarks/hpcc/stream.chpl

>  ./stream-main -v
...
Execution time:
  tot = 5.56785
  avg = 0.556785
  min = 0.555181
Performance (GB/s) = 243.05

> ./stream -v
...
Execution time:
  tot = 5.56952
  avg = 0.556952
  min = 0.555681
Performance (GB/s) = 242.831
```

**Comment by jhh67** [comment]
`no-op` results collected on an HPE Cray EX.
```
# CHPL_COMM=none, main branch

> start_test --performance test/performance/elliot/no-op.chpl
07/11/23	0.1791

> start_test --performance test/performance/elliot/no-op.chpl
07/11/23	0.1761

**Comment by jhh67** [comment]
`ig` results run on a dual-socket, dual-NIC HPE Cray EX. When run with two locales per node it was verified using `CHP_COMM_DEBUG=1`  and `CHPL_RT_COMM_OFI_DEBUG=prov` that the locales used the NIC in their socket. Note that in the output GB/s/node is actually GB/s/locale.
```
> chpl --fast test/studies/bale/aggregation/ig.chpl -sN=10000000 -suseBlockArr=true

# main branch

> ./ig-main -nl 16
       AGG:	1.739 seconds	11.778 GB/s/node

> env CHPL_RT_LOCALES_PER_NODE=2 ./ig-main -nl 32
       AGG:	1.419 seconds	7.217 GB/s/node

# this PR

> ./ig-pr -nl 16
       AGG:	1.732 seconds	11.824 GB/s/node

> env CHPL_RT_LOCALES_PER_NODE=2 ./ig-pr -nl 32
       AGG:	1.435 seconds	7.134 GB/s/node

File: third-party/libfabric/Makefile (hunk 4 above)
Comment (jhh67) [reply third-party/libfabric/Makefile]: Undone.

File: util/chplenv/third_party_utils.py (hunk 5 above)
Comment (jhh67) [reply util/chplenv/third_party_utils.py:199]: Removed this code and modified `dedup` to only dedup `-l` args. That was the least complex solution.

File: util/chplenv/chpl_hwloc.py [hunk 6]
```diff
@@ -46,10 +47,9 @@ def get_link_args():
...
     elif hwloc_val == 'system':
         # Check that hwloc version is OK
-        okversions = ('1.11.5', '1.11.6', '1.11.7', '1.11.8', '1.11.9', '1.11.10', '1.11.11', '1.11.12', '1.11.13')
-        vers = third_party_utils.pkgconfig_get_system_version('hwloc')
-        if not vers in okversions:
-          err = "CHPL_HWLOC=system but unsupported version {0} was found.\nPlease use one of the following versions {1}\n".format(vers, ' '.join(okversions))
+        sub = subprocess.run(['pkg-config', '--exists', 'hwloc >= 2.0'])
+        if sub.returncode != 0:
+          err = "CHPL_HWLOC=system requires hwloc >= 2.0"
           error(err, ValueError)
```
[the remark this answers is not in the document]
Comment (jhh67) [reply util/chplenv/chpl_hwloc.py:54]: I've made the modifications you recommended. I left in the version check as I prefer to catch such errors at build time rather than runtime.

File: runtime/src/topo/hwloc/topo-hwloc.c (hunk 1 above)
Comment (jhh67) [reply runtime/src/topo/hwloc/topo-hwloc.c:157]: The runtime won't compile with `hwloc` < 2.1, so it will never get to the runtime check so I removed it.

File: third-party/hwloc/README (hunk 3 above)
Comment (jhh67) [reply third-party/hwloc/README:33]: Done.

Comment (ronawho) [reply third-party/hwloc/README:32]: Paragraph still has some text about it.

File: util/chplenv/chpl_hwloc.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 import sys
 
 import chpl_locale_model, chpl_tasks, overrides, third_party_utils
+import subprocess
```
Comment (ronawho) [inline util/chplenv/chpl_hwloc.py:5]: Unnecessary subprocess import

File: third-party/hwloc/README (hunk 2 above)
Comment (jhh67) [reply third-party/hwloc/README:39]: I've created an issue (https://github.com/Cray/chapel-private/issues/5102) to do this.

File: third-party/hwloc/README (hunk 3 above)
Comment (jhh67) [reply third-party/hwloc/README]: Fixed.

**Review by ronawho — APPROVED** [review]

**Comment by bradcray** [comment]
@jhh67 / @ronawho : Does this merge include the fix that would permit us to use Qthreads on ARM?

**Comment by ronawho** [comment]
No, this is hwloc only. qthreads on arm requires a qthreads upgrade.

**Comment by bradcray** [comment]
Oh my gosh, my brain is still on vacation.  Let's just pretend I never asked this (where's that delete button...).

(I guess in my defense, hwloc and qthreads tend to be pretty intertwined relative to our other third-party stuff...)
